src/Evolution/Systems/GeneralizedHarmonic/UpwindPenaltyCorrection.hpp

```diff
@@ -0,0 +1,236 @@
+ *              &-\frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}_{ab} +
+ *                \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}_{ab}\right)
+ *                - \tilde{\lambda}_{v^g}^\mathrm{int}\gamma_2
+ *                v^{\mathrm{int},g}_{ab} , \\
+ *   \label{eq:pnpm upwind gh phi}
+ *   D_{\Phi_{iab}}
+ *              &= \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},+}_{ab}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},-}_{ab}\right)n_i^{\mathrm{ext}}
+ *                + \tilde{\lambda}_{v^0}^{\mathrm{ext}}
+ *                v^{\mathrm{ext},0}_{iab}
+ *                \notag \\
+ *              &-
+ *                \frac{1}{2}\left(\tilde{\lambda}_{v^+}^{\mathrm{int}}
+ *                v^{\mathrm{int},+}_{ab}
+ *                - \tilde{\lambda}_{v^-}^{\mathrm{int}}
+ *                v^{\mathrm{int},-}_{ab}\right)n_i^{\mathrm{int}}
+ *                - \tilde{\lambda}_{v^0}^{\mathrm{int}}
+ *                v^{\mathrm{int},0}_{iab},
+ * \f}
+ *
+ * with characteristic fields
+ *
+ * \f{align}{
+ *   \label{eq:pnpm gh v g}
+ *   v^{g}_{ab} &= g_{ab}, \\
+ *   \label{eq:pnpm gh v 0}
+ *   v^{0}_{iab} &= (\delta^k_i-n^k n_i)\Phi_{kab}, \\
+ *   \label{eq:pnpm gh v pm}
+ *   v^{\pm}_{ab} &= \Pi_{ab}\pm n^i\Phi_{iab} -\gamma_2 g_{ab},
+ * \f}
+ *
+ * and characteristic speeds
+ *
+ * \f{align}{
+ *   \lambda_{v^g} =& -(1+\gamma_1)\beta^i n_i -v^i_g n_i, \\
+ *   \lambda_{v^0} =& -\beta^i n_i -v^i_g n_i, \\
+ *   \lambda_{v^\pm} =& \pm \alpha - \beta^i n_i - v^i_g n_i,
+ * \f}
+ *
+ * where \f$v_g^i\f$ is the mesh velocity. We have also defined
+ *
+ * \f{align}{
+ *   \tilde{\lambda}_{\hat{\alpha}} =
+ *    \left\{
+ *        \begin{array}{ll}
+ *          \lambda_{\hat{\alpha}} &
+ *            \mathrm{if}\;\lambda_{\hat{\alpha}}\le 0.0 \\
+ *          0 & \mathrm{otherwise}
+ *        \end{array}\right.
+ * \f}
+ *
+ * Note that we have assumed \f$n_i^{\mathrm{ext}}\f$ points in the same
+ * direction as \f$n_i^{\mathrm{int}}\f$. If \f$n_i^{\mathrm{ext}}\f$ points in
+ * the opposite direction the external speeds have their sign flipped and the
+ * \f$\pm\f$ fields and speeds reverse roles.
+ */
+template <size_t Dim>
+struct UpwindPenaltyCorrection : tt::ConformsTo<dg::protocols::NumericalFlux> {
+  struct NormalTimesVPlus : db::SimpleTag {
+    using type = tnsr::iaa<DataVector, Dim, Frame::Inertial>;
+  };
+  struct NormalTimesVMinus : db::SimpleTag {
+    using type = tnsr::iaa<DataVector, Dim, Frame::Inertial>;
+  };
+  struct Gamma2VSpacetimeMetric : db::SimpleTag {
+    using type = tnsr::aa<DataVector, Dim, Frame::Inertial>;
+  };
+  struct TensorCharSpeeds : db::SimpleTag {
+    // use a tensor for sending the char speeds so they all go into a Variables
+    // to reduce memory allocations.
+    using type = tnsr::a<DataVector, 3, Frame::Inertial>;
+  };
+
```

Comment:
I presume the tags inside the class mean they are private to the class...perhaps file an issue to replace char speeds with tensor char speeds everywhere, since they are computed a lot of places, mostly as `std::arrays`? I understand this would be outside the scope of this PR.

Reply:
I don't remember why exactly we use a `std::array` for the char speeds, but there was a lot of back and forth. Going forward I'm planning on having the boundary terms/numerical fluxes compute the char speeds internally so it should be less of a problem, but I'll file an issue to revisit once we know what's happening.